lib/ffmpeg/transcoderbase.py

```diff
@@ -227,67 +238,75 @@ def transcode(self):
-
-          assert (len(input_mdm_info) == 0 or (len(input_mdm_info) == len(output_mdm_info) and input_mdm_info[0] == output_mdm_info[0])) and \
-            (len(input_cll_info) == 0 or (len(input_cll_info) == len(output_cll_info) and input_cll_info[0] == output_cll_info[0])), "HDR info is different between input and output"
 
-        # delete yuv file after each iteration
-        get_media().artifacts.purge(yuv)
+        self.check_resolution(output, osencoded)
+        self.check_hdr(output, osencoded)
+        self.check_metrics(output, osencoded, self.goutputs[f"{n}_ref"], refctx = [(n, channel)])
 
   def check_resolution(self, output, encoded):
     actual = ffmpeg_probe_resolution(encoded)
     expect = "{}x{}".format(
       output.get("width", self.width), output.get("height", self.height))
     assert expect == actual
 
-  def check_metrics(self, yuv, refctx):
+  def check_hdr(self, output, encoded):
+    if output.get("tonemap", None) is not None:
+      get_media()._set_test_details(**{"hdr.info:expect" : "no"})
+      # make sure the output is a SDR file
+      output_mdm_info, output_cll_info = self.get_hdr_info(encoded)
+      assert len(output_mdm_info) == 0 and len(output_cll_info) == 0, "Found HDR information in output video, h2s failed"
+    elif vars(self).get("hdr", False):
+      get_media()._set_test_details(**{"hdr.info:expect" : "yes"})
+      input_mdm_info, input_cll_info = self.get_hdr_info(filepath2os(self.source))
+      assert len(input_mdm_info) + len(input_cll_info) > 0, "Found no HDR information in input video"
+
+      output_mdm_info, output_cll_info = self.get_hdr_info(encoded)
+      assert len(output_mdm_info) + len(output_cll_info) > 0, "Found no HDR information in output video"
+
+      assert (len(input_mdm_info) == 0 or (len(input_mdm_info) == len(output_mdm_info) and input_mdm_info[0] == output_mdm_info[0])) and \
+        (len(input_cll_info) == 0 or (len(input_cll_info) == len(output_cll_info) and input_cll_info[0] == output_cll_info[0])), "HDR info is different between input and output"
+
+  def check_metrics(self, output, encoded, reference, refctx):
+    iopts   = ""
+    uformat = output.get("format", self.format) # unmapped output format
+    oformat = self.map_format(uformat) # mapped output format
+    ocodec  = output["codec"]
+
+    # WA: FFMpeg does not have an AV1 SW decoder
+    if ocodec in [Codec.AV1]:
+      iopts = (
+        f"-hwaccel {self.hwaccel}"
+        f" -init_hw_device {self.hwaccel}={self.hwdevice}"
+        f" -hwaccel_output_format {oformat}"
+        f" -c:v {self.get_decoder(ocodec, 'hw')}"
+      )
+
+    iopts += f" -i {encoded}"
+
+    yuv = get_media().artifacts.reserve("yuv")
+    osyuv = filepath2os(yuv)
+    vppscale = self.get_vpp_scale(self.width, self.height, "sw")
+    oopts = (
+      f"-vf '{vppscale}' -pix_fmt {oformat} -f rawvideo"
+      f" -vframes {self.frames} -y {osyuv}"
+    )
+
+    self.call_ffmpeg(iopts, oopts)
```

Comment:
This second run is the intermediate step that dumps the xcode output result to "yuv" so that it can be compared with the "reference" at the next line (`metrics.check...`).  It is the same code (slightly improved) that was moved out of the inner-loop (channels) in the "transcode" method.